src/mds/MDCache.cc

```diff
@@ -13448,6 +13504,282 @@ void MDCache::clear_dirty_bits_for_stray(CInode* diri) {
+    dout(20) << __func__ << " " << *mdr << " already finished quiesce" << dendl;
+    return;
+  }
+
+  auto* qfinisher = static_cast<C_MDS_QuiesceSubvolume*>(mdr->internal_op_private);
+  auto delay = qfinisher->delay;
+  auto splitauth = qfinisher->splitauth;
+  auto& qs = *qfinisher->qs;
+  auto qrmdr = qfinisher->mdr;
+
+  CInode *in = get_inode(mdr->get_filepath().get_ino());
+  if (in == nullptr) {
+    qs.add_failed(mdr, -ENOENT);
+    mds->server->respond_to_request(mdr, -ENOENT);
+    return;
+  }
+  const bool is_root = (mdr->get_filepath().get_ino() == mdr->get_filepath2().get_ino());
+
+  dout(20) << __func__ << " " << *mdr << " quiescing " << *in << dendl;
+
+  /* Acquire rdlocks on anything which prevents writing.
+   *
+   * Because files are treated specially allowing multiple reader/writers, we
+   * need an xlock here to recall all write caps. This unfortunately means
+   * there can be no readers.
+   *
+   * The xlock on the quiescelock is important to prevent future requests
+   * from blocking on other inode locks while holding path traversal locks.
+   * See dev doc doc/dev/mds_internals/quiesce.rst for more details.
+   */
+
+  MutationImpl::LockOpVec lov;
+  if (!mds->locker->acquire_locks(mdr, lov, nullptr, {in}, false, true)) {
+    return;
+  }
+  /* TODO: Consider:
+   *
+   *  rank0 is auth for /foo
+   *  rank1 quiesces /foo with no dirents in cache (and stops)
+   *  rank0 begins quiescing /foo
+   *  rank0 exports a dirfrag of /foo/bar to rank1 (/foo/bar is not authpinned by rank1 nor by rank0 (yet))
+   *  rank1 discovers relevant paths in /foo/bar
+   *  rank1 now has /foo/bar in cache and may issue caps / execute operations
+   *
+   * The solution is probably to have rank1 mark /foo has STATE_QUIESCED and reject export ops from rank0.
+   */
+
+  if (in->is_auth()) {
+    MutationImpl::LockOpVec lov;
+    lov.add_rdlock(&in->authlock);
+    lov.add_rdlock(&in->dirfragtreelock);
+    if (in->is_file()) {
+      lov.add_xlock(&in->filelock);
+    } else {
+      lov.add_rdlock(&in->filelock);
+    }
+    lov.add_rdlock(&in->linklock);
+    lov.add_rdlock(&in->nestlock);
+    lov.add_rdlock(&in->policylock);
+    // N.B.: NO xlock/wrlock on quiescelock; we need to allow access to mksnap/lookup
+    // This is an unfortunate inconsistency. It may be possible to circumvent
+    // this issue by having those ops acquire the quiscelock only if necessary.
+    if (is_root) {
+      lov.add_rdlock(&in->quiescelock);
+    } else {
+      lov.add_xlock(&in->quiescelock); /* !! */
+    }
+    lov.add_rdlock(&in->snaplock);
+    lov.add_rdlock(&in->xattrlock);
+    if (!mds->locker->acquire_locks(mdr, lov, nullptr, {in}, false, true)) {
+      return;
+    }
+  } else if (!splitauth) {
+    dout(5) << "auth is split and splitauth is false: " << *in << dendl;
+    qs.add_failed(mdr, -EPERM);
+    mds->server->respond_to_request(mdr, -EPERM);
+    return;
+  }
+
+  if (in->get_projected_inode()->get_quiesce_block()) {
```

Comment:
Given the expense of quiescing a tree in general, this is not really significant.

Reply:
I thought it could be used to prevent deadlocks. Maybe i want to exclude this from quiesce because my app claims a permanent exclusive lock on it

Reply:
we could make this check after only taking the `xattrlock` if we wanted to make this lighter for any potentially outstanding write capabilities. This will be significant for single writer client with unflushed buffers

Reply:
another thing, this should aways call `in->get_inode()`, and whether it's the projected one or the current one will be decided at the iteration level, where we'll have to spawn `quiesce_inode` for both the projected and the current inodes.

Reply:
> I thought it could be used to prevent deadlocks. Maybe i want to exclude this from quiesce because my app claims a permanent exclusive lock on it

That is not possible in CephFS. Capabilities are always subject to recall.

Reply:
> we could make this check after only taking the `xattrlock` if we wanted to make this lighter for any potentially outstanding write capabilities. This will be significant for single writer client with unflushed buffers

That would complicate lock order requirements. It really is not necessary.

Reply:
> That is not possible in CephFS. Capabilities are always subject to recall.

I agree it can't cause a deadlock this way, but it will at least cause unnecessary buffer flushes for clients who may want to avoid it by setting this flag

> That would complicate lock order requirements. It really is not necessary.

I believe it is not dangerous because
* we take rdlocks
* any other operation that attempts a wrlock on either of the four cap locks will be forced to also take a rdlock on the quiesce, and that we already hold xlocked

Reply:
> > That is not possible in CephFS. Capabilities are always subject to recall.
> 
> I agree it can't cause a deadlock this way, but it will at least cause unnecessary buffer flushes for clients who may want to avoid it by setting this flag

Huh? We want clients to flush buffers before an inode is quiesced?

> > That would complicate lock order requirements. It really is not necessary.
> 
> I believe it is not dangerous because
> 
>     * we take rdlocks
> 
>     * any other operation that attempts a wrlock on either of the four cap locks will be forced to also take a rdlock on the quiesce, and that we already hold xlocked

Taking locks out of order should be verboten no matter the circumstances. It is a recipe for future bugs.

I must emphasize that you're proposing an **optimization** for a code path that won't ever be taken with the way quiesce will work in production for subvolumes. We are not quiescing the `.meta` file at all anymore. I don't think this is worth the trouble.

Reply:
> Huh? We want clients to flush buffers before an inode is quiesced?

*Clients* want to flush buffers for ordered writes. We don't care if a client app doesn't flush for a crash-safe checkpoint. This was a big point in the beginning of our presentation: the contract only works if both sides adhere to the rules. Unflushed writes may be staying in the OS buffers so even if we revoke the GBUFFER from a client it still doesn't mean that the writes from the apps are fully stopped.

So, we don't care about those buffers. Apps should care and they will call flush regularly, which is what we will have in the snapshots.

Reply:
> I must emphasize that you're proposing an optimization for a code path that won't ever be taken with the way quiesce will work in production for subvolumes. We are not quiescing the .meta file at all anymore. I don't think this is worth the trouble.

I agree that we no longer need that flag for our implementation.

However, if we do want to stay with the flag as a feature, I believe we should do what we can to avoid revoking caps for a dir that was marked with this flag. That would be a proper way of implementing it.

Plus, as I said, even though dangerous, we can afford id here because the quiesce lock is the first lock, and it is so by definition. As long as we have that lock xlocked, and other locks that we play with will rdlock it, it's safe to take them in any order